lib/Sema/TypeCheckType.cpp

```diff
@@ -1212,6 +1212,56 @@ static Type diagnoseUnknownType(TypeResolution resolution,
   return ErrorType::get(ctx);
 }
 
+static bool shouldPreferModulesForTopLevelIdentTypeComponent(
+    const DeclContext *DC, TypeResolutionOptions options) {
+
+  // FIXME: We have a problem when a module and a type have the same name; the
+  // underlying UnqualifiedLookup code assumes that the type is always what's
+  // meant, even when that might not be the case. Work around this for
+  // parseable interfaces by making the *opposite* assumption outside of
+  // expressions and function bodies.
+  const SourceFile *containingSF = DC->getParentSourceFile();
+  if (!containingSF || containingSF->Kind != SourceFileKind::Interface)
+    return false;
+
+  // If this is not a local context, just check whether we're /in/ a local
+  // context.
+  if (!DC->isLocalContext())
+    return DC->getLocalContext() == nullptr;
+
+  // We still want to return true for function parameters and result types, as
+  // long as they're not in *another* local context.
```

Comment:
Inlinable functions cannot contain nested types so I think some of these checks could be simplified.

Reply:
I don't think it gets any simpler than what's here. Inlinable functions can still contain other functions.

Reply:
(and typealiases)

Reply:
When checking parameters and results are we in the DC of the function or are we in the parent DC?

Reply:
The function, because we need to find the function's generic parameters.